File: Scala/string_algorithms/src/main/scala/burrows_wheeler/SuffixArray.scala
package burrows_wheeler

object SuffixArray {
  import scala.annotation.tailrec

  def calcSuffixArray(text: String): List[Int] = {
    val length: Int = text.length

    def compare(i: Int, j: Int): Boolean = {
      val limit: Int = math.min(length - i, length - j)
      @tailrec
      def loop(k: Int): Boolean = {
        if (k == limit) false
        else if (text(i + k) < text(j + k)) true
        else if (text(i + k) > text(j + k)) false
        else loop(k + 1)
      }
      loop(0)
    }

    text.indices.sortWith(compare).toList
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val text: String = reader.next()
    val result: List[Int] = calcSuffixArray(text)
    println(result.mkString(" "))
  }
}

File: OCaml/string_algorithms/burrows_wheeler/suffix_array.ml
open Batteries

let calc_suffix_array text =
    let length = String.length text in
    let compare i j =
        let limit = min (length - i) (length - j) in
        let rec loop k =
            if k = limit then 1
            else if String.get text (i + k) < String.get text (j + k) then -1
            else if String.get text (i + k) > String.get text (j + k) then 1
            else loop (k + 1)
        in loop 0
    in List.sort compare @@ List.init length identity


let () =
    let text = read_line() in
    let result = calc_suffix_array text in
    List.iter (function elem -> print_int elem; print_string " ") result;
    print_newline()
